jupyterhub/handlers/base.py

```diff
@@ -1554,7 +1564,19 @@ def get(self):
             # default / -> /hub/ redirect
             # avoiding extra hop through /hub
             path = '/'
-        self.redirect(url_path_join(self.hub.base_url, path), permanent=False)
+
+        redirect_url = redirect_path = url_path_join(self.hub.base_url, path)
+
+        # when using subdomains,
+        # make sure we redirect `user.domain/user/foo` -> `hub.domain/hub/user/foo/...`
+        # so that the Hub handles it properly with cookies and all
+        public_url = self.settings.get("public_url")
+        subdomain_host = self.settings.get("subdomain_host")
+        if public_url:
+            redirect_url = urlunparse(public_url._replace(path=redirect_path))
+        elif subdomain_host:
+            redirect_url = url_path_join(subdomain_host, redirect_path)
```

Comment:
is this right? a few things

- if subdomain_host is set, public_url also gets a default value
- subdomain host may be different than where the host lives?

Reply:
I guess this is right. Not sure why I'm not seeing the redirect I expect in my deployment.

Reply:
Ok, I've verified this fix!

Reply:
Thanks for confirming! Yes, the hub ought to be at `subdomain_host`. It is valid and reasonable for the Hub to be at a different domain, or its own subdomain peer to users (technically this is possible now because subdomain_hook can ignore subdomain_host for user URLs), but as it currently works `subdomain_host == hub host`.